Create rule S6402[terraform] Creating DNS zones without DNSSEC enabled is security-sensitive (#691)
SonarSource/rspec: rules/S6402/recommended.adoc

== Recommended Secure Coding Practices

> It's recommended to use DNSSEC when creating private and public DNS zones. Private DNS zones cannot be queried on the Internet and provide DNS name resolution for private networks. Private networks might be considered as trusted and may not immediately require DNSSEC. Note that when choosing a cryptographic DNSSEC algorithm (`default_key_specs.algorithm` argument) `rsasha1` should not be used as it's not anymore considered as a https://en.wikipedia.org/wiki/SHA-1[secure algorithm].

Reviewer: For this paragraph, I also suggest:
* breaking it into multiple paragraphs, to ease readability.
* And I think that the crypto part makes sense, and at the same time, I think it could be removed as it's out-of-scope. However, we should keep it in mind for the future encryption sprint

```suggestion
It is recommended to use DNSSEC when creating private and public DNS zones. +
Private DNS zones cannot be queried on the Internet and provide DNS name resolution for private networks: They might be considered trusted and may not immediately require DNSSEC. 
```

Reviewer 2: I don't think we need a rule about crypto dnssec algorithms, at least I don't plan to create it during this sprint.

Just some explanations:
- `rsasha1` cannot be used freely in GCP, an error will be produced and outputted by terraform if you do that, a prior and explicit request to Google is necessary and in this request the devops should explain to Google why he wants to use this algorithm (likely for backward compatibility). So I don't think it provides value to highlight `rsasha1` as the reason of using it has been already explained and justified by the user and validated by Google/GCP.
- Let's suppose there is no restriction about using `rsasha1` then without indicating what algorithms to use, the following scenario will be encountered:
   - The user creates a DNS zone without DNSSEC in a first commit of a PR  
   - SonarQube raises a S6402 issue
   - The user push a commit to fix this issue, without indication in the s6402 description, the user could choose a weak algorithm (`rsasha1`).
   - SonarQube raises a SXXXX issue because the user has to choose a strong algorithm
   - The user has to push a second commit to fix this issue

By experience this user experience is not really good and could be simply avoided by providing the necessary information in the rule description.

WDYT?

Reviewer: Two points: Yes, creating a DNSSEC algorithm during the current sprint is out-of-scope; that's why I precise we should keep it in mind for the future encryption sprint.

Yes, I agree that avoiding too much context-switching to fix things. So let's do it.

It's strange that terraform mentions RSA+SHA1 since GCP doesn't  seem to enable it:
![image](https://user-images.githubusercontent.com/91723853/149517283-e7b46984-db69-4fd5-bd2a-6cac816529e2.png)
[LINK](https://cloud.google.com/dns/docs/dnssec-advanced)

For this suggestion, I'm not sure whether to suggest we use a more imperative voice, and say: "Use a strong algorithm, like RSASHA256", instead of saying "don't use RSASHA1". We should either mention the strong algorithms or link to the docs listing the strong algorithms IMO

I think I would suggest something along those lines then:
```suggestion
It is recommended to use DNSSEC when creating private and public DNS zones. +
Private DNS zones cannot be queried on the Internet and provide DNS name resolution for private networks: They might be considered trusted and may not immediately require DNSSEC. 

Note: Choose a robust signing algorithm when setting up DNSSEC, like `rsasha256`. Other algorithms like `rsasha1` should not be used as it is now considered an https://en.wikipedia.org/wiki/SHA-1[insecure algorithm].
```

Reviewer 2: I like you suggestion!

I saw this page but according to various SO posts, `rsasha1` can apparently still be used:
https://stackoverflow.com/questions/68968312/terraform-errors-deploying-google-dns-managed-zone-with-rsasha1
but of course I cannot confirm

Reviewer: Ah ok, so if a customer really wants to have SHA1 enabled they have to contact GCP support haha. I wonder what cases require SHA-1
